runtime-transaction/src/compute_budget_instruction_details.rs

```diff
@@ -0,0 +1,349 @@
+use {
+    solana_compute_budget::compute_budget_limits::*,
+    solana_sdk::{
+        borsh1::try_from_slice_unchecked,
+        compute_budget::{self, ComputeBudgetInstruction},
+        instruction::{CompiledInstruction, InstructionError},
+        pubkey::Pubkey,
+        saturating_add_assign,
+        transaction::{Result, TransactionError},
+    },
+};
+
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Default, Debug)]
+pub struct ComputeBudgetInstructionDetails {
+    // compute-budget instruction details:
+    // the first field in tuple is instruction index, second field is the unsanitized value set by user
+    pub requested_compute_unit_limit: Option<(u8, u32)>,
+    pub requested_compute_unit_price: Option<(u8, u64)>,
+    pub requested_heap_size: Option<(u8, u32)>,
+    pub requested_loaded_accounts_data_size_limit: Option<(u8, u32)>,
+    pub count_compute_budget_instructions: u32,
+}
+
+impl ComputeBudgetInstructionDetails {
+    pub fn process_instruction<'a>(
+        &mut self,
+        index: u8,
+        program_id: &'a Pubkey,
+        instruction: &'a CompiledInstruction,
+    ) -> Result<()> {
+        if compute_budget::check_id(program_id) {
+            let invalid_instruction_data_error =
+                TransactionError::InstructionError(index, InstructionError::InvalidInstructionData);
+            let duplicate_instruction_error = TransactionError::DuplicateInstruction(index);
+
+            match try_from_slice_unchecked(&instruction.data) {
+                Ok(ComputeBudgetInstruction::RequestHeapFrame(bytes)) => {
+                    if self.requested_heap_size.is_some() {
+                        return Err(duplicate_instruction_error);
+                    }
+                    if Self::sanitize_requested_heap_size(bytes) {
```

Comment:
Shouldn't this be done in the sanitize method?

Reply:
Been back-n-forth on this one, should error out bad value at instruction processing or during sanitization? After making raw structs private, it's clearer to me it should be done in sanitize. Thanks for pointing it out.

Reply:
f22680528da144b7c94c8dc81c991310c49cd33d